README.md
# sitemap-xmls

`sitemap-xmls` is a high-level sitemap-generating framework that makes creating sitemap XML files easy.

## Installation

```bash
npm install sitemap-xmls --save
```

## Usage

```js
const smXml = require('sitemap-xmls');

const sitemapXml = new smXml('demo', 'https://demo.com/sitemap/',{
  releasePath: './release/'
});

// add a sitemap with type of 'apps'
sitemapXml.addSitemapType('apps', 'monthly', 0.8);

sitemapXml.addLine('apps', 'https://demo.com/apps-demo-1.html');
sitemapXml.addLine('apps', 'https://demo.com/apps-demo-2.html');
sitemapXml.addLine('apps', 'https://demo.com/apps-demo-3.html', {
  lastmod: '2018-10-01',
  changefreq: 'daily',
  priority: 0.6
});

// important: don't forget the 'end()'
sitemapXml.end('apps');


// add another sitemap with type of 'topics'
sitemapXml.addSitemapType('topics', 'daily', 0.9);

sitemapXml.addLine('topics', 'https://demo.com/topics-demo-1.html');
sitemapXml.addLine('topics', 'https://demo.com/topics-demo-2.html');
sitemapXml.addLine('topics', 'https://demo.com/topics-demo-3.html');

// important: don't forget the 'end()'
sitemapXml.end('topics');


// finally, generate the index file
sitemapXml.createIndex();
```

after that, it will generator sitemap just like in `./release/demo/`:

- `demo/index.json`
- `demo/index.xml`
- `demo/apps_10000.xml.gz`
- `demo/topics_10000.xml.gz`

【demo/index.json】

```json
{"index":"demo/index.xml","files":["demo/apps_10000.xml.gz","demo/topics_10000.xml.gz"]}
```

【demo/index.xml】

```xml
<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
<sitemap>
<loc>https://demo.com/sitemap/demo/apps_10000.xml.gz</loc>
<loc>https://demo.com/sitemap/demo/topics_10000.xml.gz</loc>
</sitemap>
</sitemapindex>
```

【demo/apps_10000.xml.gz】

```xml
<?xml version="1.0" encoding="UTF-8"?><urlset>
<url>
<loc>https://demo.com/apps-demo-1.html</loc>
<lastmod>2018-07-18</lastmod>
<changefreq>monthly</changefreq>
<priority>0.8</priority>
</url>
<url>
<loc>https://demo.com/apps-demo-2.html</loc>
<lastmod>2018-07-18</lastmod>
<changefreq>monthly</changefreq>
<priority>0.8</priority>
</url>
<url>
<loc>https://demo.com/apps-demo-3.html</loc>
<lastmod>2018-10-01</lastmod>
<changefreq>daily</changefreq>
<priority>0.6</priority>
</url>
</urlset>
```

【demo/topics_10000.xml.gz】

```xml
<?xml version="1.0" encoding="UTF-8"?><urlset>
<url>
<loc>https://demo.com/topics-demo-1.html</loc>
<lastmod>2018-07-18</lastmod>
<changefreq>daily</changefreq>
<priority>0.9</priority>
</url>
<url>
<loc>https://demo.com/topics-demo-2.html</loc>
<lastmod>2018-07-18</lastmod>
<changefreq>daily</changefreq>
<priority>0.9</priority>
</url>
<url>
<loc>https://demo.com/topics-demo-3.html</loc>
<lastmod>2018-07-18</lastmod>
<changefreq>daily</changefreq>
<priority>0.9</priority>
</url>
</urlset>
```

## Params

### changefreq

- always
- hourly
- daily
- weekly
- monthly
- yearly
- never

### priority

0.0 ~ 1.0

### mobile

```xml
<mobile:mobile type="mobile"/> ：移动网页      
<mobile:mobile type="pc,mobile"/>：自适应网页
<mobile:mobile type="htmladapt"/>：代码适配
```

无该上述 `mobile` 标签表示为PC网页。

## 附：相关文档

- [百度 sitemap](https://ziyuan.baidu.com/college/courseinfo?id=267&page=2#04)
- [sitemaps.org](https://www.sitemaps.org/index.html)


History.md

1.4.n / 2018-09-12
==================

  * feat: SitemapXml 构造函数增加changefreq等三个参数的支持

1.3.0 / 2018-09-11
==================

  * feat: lastmod 不再默认指定，由外面传进来

1.2.1 / 2018-08-08
==================

  * fix: 解决google命名空间问题

1.2.0 / 2018-07-26
==================

  * feat: 几个配置不配的话 sitemap 上可为空
  * Update README.md

1.1.2 / 2018-07-21
==================

  * feat: 生成索引文件的同时，多生成一份包含所有文件列表的json文件

1.1.1 / 2018-07-19
==================

  * feat: 索引文件添加 lastmod

1.1.0 / 2018-07-19
==================

  * feat: 写URL时增加参数传递，另外增加参数 mobile 的支持

1.0.4 / 2018-07-19
==================

  * fix: 修复URL少了斜杠 '/' 的问题
  * deps: add mkdirp to depenencies

1.0.3 / 2018-07-19
==================

  * deps: add mkdirp to depenencies

1.0.2 / 2018-07-18
==================

  * feat: 文件写入完成时添加回调
  * fix: 修复 utils
  * Update README.md

1.0.1 / 2018-07-18
==================

  * fix: 修复日期问题
  * doc: update readme.md
  * feat: add History.md

1.0.0 / 2018-07-18
==================

  * feat: project init
  * Initial commit


lib/utils.js
/**
 * utils 通用方法
 *
 * @author 材主<diamont1001@163.com>
 */

'use strict';

class Utils {

  static insertArray(arr, i, item) {
    arr.splice(i, 0, item);
  }

  // 判断是否对象，并且为非空和非 {}
  static isObjectNotEmpty(a) {
    if (typeof a === 'object' && !(a instanceof Array)) {
      for (const i in a) {
        return true;
      }
      return false;
    }
    return false;

  }

  /**
   * 时间戳格式化函数
   * @param  {string} format  格式 如'Y-m-d H:i:s'
   *   Year -
   *     L: 闰年（0/1）
   *     Y: 年份（数字4位），比如：2016
   *     y: 年份（数字2位），比如：16
   *
   *   Month -
   *     F: 月份（英文全写），比如：January
   *     M: 月份（英文缩写3个字母），比如：Sep
   *     m: 月份（01 ~ 12）
   *     n: 月份（1 ~ 12）
   *     t: 该月拥有天数（28 ~ 31）
   *
   *   Week -
   *     W: 第几个星期（0~52）
   *
   *   Day -
   *     d: 日期-天（01 ~ 31）
   *     j: 日期-天（1 ~ 31）
   *     S: 日期-天的英文后缀，比如：th, nd, st, rd, st
   *     l: 星期几（英文全写），比如：Tuesday
   *     D: 星期几（英文缩写3个字母），比如：Thu
   *     w: 一周的第几天（0 ~ 6）
   *     N: 一周的第几天（1 ~ 7）
   *     z: 一年的第几天（0 ~ 365）
   *
   *   Time -
   *     a: am / pm
   *     A: AM / PM
   *     H: 小时（00 ~ 23）
   *     h: 小时（01 ~ 12）
   *     G: 小时（0 ~ 23）
   *     g: 小时（1 ~ 12）
   *     i: 分钟（00 ~ 59）
   *     s: 秒（00 ~ 59）
   *     B: 未知
   *
   *   Timezone(时区) -
   *     O: +数字，比如：正八区为 +0800
   *     P: +数字:数字，比如：正八区为 +08:00
   *
   *   other -
   *     c: Full Date/Time，比如：2006-12-31T12:13:00+08:00
   *     U: 时间戳（秒）
   *
   * @param  {int}  timestamp 要格式化的时间戳 默认为当前时间
   * @return {string}      格式化后的时间字符串
   */
  static stampFormat2Date(format, timestamp) {
    timestamp = timestamp <= 9999999999 ? timestamp * 1000 : timestamp; // 兼容传进来timestamp的单位是s的情况

    var jsdate = ((timestamp) ? new Date(timestamp) : new Date());
    var pad = function(n, c) {
      if ((n = n + '').length < c) {
        return new Array(++c - n.length).join('0') + n;
      } else {
        return n;
      }
    };
    var txt_weekdays = [ 'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday' ];
    var txt_ordin = {
      1: 'st',
      2: 'nd',
      3: 'rd',
      21: 'st',
      22: 'nd',
      23: 'rd',
      31: 'st'
    };
    var txt_months = [ '', 'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December' ];
    var f = {
      // Day
      d: function() {
        return pad(f.j(), 2);
      },
      D: function() {
        return f.l().substr(0, 3);
      },
      j: function() {
        return jsdate.getDate();
      },
      l: function() {
        return txt_weekdays[f.w()];
      },
      N: function() {
        return f.w() + 1;
      },
      S: function() {
        return txt_ordin[f.j()] ? txt_ordin[f.j()] : 'th';
      },
      w: function() {
        return jsdate.getDay();
      },
      z: function() {
        return (jsdate - new Date(jsdate.getFullYear() + '/1/1')) / 864e5 >> 0; // eslint-disable-line no-bitwise
      },

      // Week
      W: function() {
        var a = f.z(),
          b = 364 + f.L() - a;
        var nd2,
          nd = (new Date(jsdate.getFullYear() + '/1/1').getDay() | 7) - 1;
        if (b <= 2 && ((jsdate.getDay() | 7) - 1) <= 2 - b) {
          return 1;
        } else {
          if (a <= 2 && nd >= 4 && a >= (6 - nd)) {
            nd2 = new Date(jsdate.getFullYear() - 1 + '/12/31');
            return Utils.stampFormat2Date('W', Math.round(nd2.getTime() / 1000));
          } else {
            return (1 + (nd <= 3 ? ((a + nd) / 7) : (a - (7 - nd)) / 7) >> 0); // eslint-disable-line no-bitwise
          }
        }
      },

      // Month
      F: function() {
        return txt_months[f.n()];
      },
      m: function() {
        return pad(f.n(), 2);
      },
      M: function() {
        return f.F().substr(0, 3);
      },
      n: function() {
        return jsdate.getMonth() + 1;
      },
      t: function() {
        var n;
        if ((n = jsdate.getMonth() + 1) === 2) {
          return 28 + f.L();
        } else {
          if (n & 1 && n < 8 || !(n & 1) && n > 7) {
            return 31;
          } else {
            return 30;
          }
        }
      },

      // Year
      L: function() {
        var y = f.Y();
        return ((y % 4 === 0) && (y % 100 !== 0 || y % 400 === 0)) ? 1 : 0; // 四年一闰，百年不闰，四百年再闰
      },
      // o not supported yet
      Y: function() {
        return jsdate.getFullYear();
      },
      y: function() {
        return (jsdate.getFullYear() + '').slice(2);
      },

      // Time
      a: function() {
        return jsdate.getHours() > 11 ? 'pm' : 'am';
      },
      A: function() {
        return f.a().toUpperCase();
      },
      B: function() {
        // peter paul koch:
        var off = (jsdate.getTimezoneOffset() + 60) * 60;
        var theSeconds = (jsdate.getHours() * 3600) + (jsdate.getMinutes() * 60) + jsdate.getSeconds() + off;
        var beat = Math.floor(theSeconds / 86.4);
        if (beat > 1000) beat -= 1000;
        if (beat < 0) beat += 1000;
        if ((String(beat)).length === 1) beat = '00' + beat;
        if ((String(beat)).length === 2) beat = '0' + beat;
        return beat;
      },
      g: function() {
        return jsdate.getHours() % 12 || 12;
      },
      G: function() {
        return jsdate.getHours();
      },
      h: function() {
        return pad(f.g(), 2);
      },
      H: function() {
        return pad(jsdate.getHours(), 2);
      },
      i: function() {
        return pad(jsdate.getMinutes(), 2);
      },
      s: function() {
        return pad(jsdate.getSeconds(), 2);
      },
      // u not supported yet

      // Timezone
      // e not supported yet
      // I not supported yet
      O: function() {
        var t = pad(Math.abs(jsdate.getTimezoneOffset() / 60 * 100), 4);
        if (jsdate.getTimezoneOffset() > 0) t = '-' + t;
        else t = '+' + t;
        return t;
      },
      P: function() {
        var O = f.O();
        return (O.substr(0, 3) + ':' + O.substr(3, 2));
      },
      // T not supported yet
      // Z not supported yet

      // Full Date/Time
      c: function() {
        return f.Y() + '-' + f.m() + '-' + f.d() + 'T' + f.h() + ':' + f.i() + ':' + f.s() + f.P();
      },
      // r not supported yet
      U: function() {
        return Math.round(jsdate.getTime() / 1000);
      }
    };

    return format.replace(/[\\]?([a-zA-Z])/g, function(t, s) {
      var ret;

      if (t !== s) {
        // escaped
        ret = s;
      } else if (f[s]) {
        // a date function exists
        ret = f[s]();
      } else {
        // nothing special
        ret = s;
      }
      return ret;
    });
  }

}

module.exports = Utils;


lib/sitemap-xml.js
/**
 * XML格式地图文件
 *
 * @author 材主<diamont1001@163.com>
 */

'use strict';

const fs = require('fs');
const path = require('path');
const { URL } = require('url');
const mkdirp = require('mkdirp');
const sm = require('sitemap');
const SubSitemap = require('./sub-sitemap');

class SitemapXml {
  /**
    * 构造函数
    * @param  {String} name  地图标识，比如「wdj」（为了支持多个站点地图生成，这里需要使用name作为地图标识）
    * @param  {String} sitemapUrl 地图在目标服务器上的地址，比如'http://www.wandoujia.com/sitemap/'
    * @param  {Object} option {
    *                         releasePath: 地图保存路径
    *                         changefreq: 更新频率（always,hourly,daily,weekly,monthly,yearly,never）
    *                         priority: 权重（0.0 - 1.0）
    *                         mobile: 移动支持（pc,mobile,htmladapt）
    *                  }
    * @return {void} -
    */
  constructor(name, sitemapUrl, option) {
    this.name = name;
    this.subMaps = {};
    this.sitemapUrl = new URL(name, sitemapUrl).toString(); // 地图在目标服务器上的地址

    const releasePath = (option && option.releasePath) ? option.releasePath : './sitemap/';
    this.releasePath = path.join(releasePath, name); // 地图保存路径
    this.indexFile = path.join(this.releasePath, 'index.xml'); // 地图索引文件

    this.jsonFile = path.join(this.releasePath, 'index.json'); // 地图文件列表
    this.json = {
      index: path.join(name, 'index.xml'),
      files: []
    };

    this.changefreq = (option && option.changefreq) ? option.changefreq : undefined;
    this.priority = (option && option.priority !== undefined && option.priority !== null) ? option.priority : undefined;
    this.mobile = (option && option.mobile) ? option.mobile : undefined;

    mkdirp.sync(this.releasePath); // 生成路径
  }

  /**
    * 增加一种地图类型
    * @param {String} type 地图类型（要保证惟一性），比如应用详情页的地图类型可以设定为'apps'
    * @param {String} changefreq 更新频率（always,hourly,daily,weekly,monthly,yearly,never）
    * @param {Number} priority 权重（0-1）
    * @param {String} mobile 移动支持
    * @return {void} -
    */
  addSitemapType(type, changefreq, priority, mobile) {
    if (this.subMaps[type]) {
      return this.subMaps[type];
    }

    changefreq = changefreq ? changefreq : this.changefreq;
    priority = (priority !== undefined && priority !== null) ? priority : this.priority;
    mobile = mobile ? mobile : this.mobile;

    return this.subMaps[type] = new SubSitemap(this.releasePath, type, changefreq, priority, mobile);
  }

  addLine(type, url, option) {
    this.subMaps[type].addLine(url, option);
  }

  end(type) {
    this.subMaps[type].end();
  }

  // 生成索引文件
  createIndex() {
    const list = [];
    for (let item in this.subMaps) {
      if (item && this.subMaps.hasOwnProperty(item)) {
        const files = this.subMaps[item].getFiles();
        for (let i = 0; i < files.length; i++) {
          list.push(new URL(files[i], this.sitemapUrl).toString());
          this.json.files.push(path.join(this.name, files[i]));
        }
      }
    }

    const smi = sm.buildSitemapIndex({
      lastmodrealtime: true,
      urls: list
    });

    fs.writeFileSync(this.indexFile, smi.toString());
    fs.writeFileSync(this.jsonFile, JSON.stringify(this.json));

    return this.indexFile;
  }
}

module.exports = SitemapXml;


lib/sub-sitemap.js
/**
 * XML格式地图文件
 *
 * @author 材主<diamont1001@163.com>
 */

'use strict';

const path = require('path');
const mkdirp = require('mkdirp');
const SitemapXmlGen = require('./sitemap-xml-gen');
// const Utils = require('./utils');

// 分文件
class SubSitemap {

  /**
   * 构造函数
   * @param  {String} filePath  生成的地图文件保存路径
   * @param  {String} type  URL类型（要保证惟一性），比如应用详情页可以type=apps
   * @param {String} changefreq 更新频率（always,hourly,daily,weekly,monthly,yearly,never）
   * @param {Number} priority 权重（0.0 - 1.0）
   * @param {Number} mobile 移动支持
   * @return {void} -
   */
  constructor(filePath = './sitemap/', type = '', changefreq, priority, mobile) {
    this.type = type;
    this.filePath = filePath;

    // this.lastmod = Utils.stampFormat2Date('Y-m-d', Date.now());
    this.changefreq = changefreq;
    this.priority = priority;
    this.mobile = mobile;

    this.files = []; // 生成的分文件列表
    this.lineIndex = 0; // 目前所在URL行数
    this.lineLimit = 50000; // 每个文件不超过 50000 条URL，不超过 10M 大小

    mkdirp.sync(this.filePath); // 生成路径
    this.newXml(); // 建立一个新文件
  }

  newXml() {
    // 结束现有的
    this.end();

    let file = (10000 + this.files.length) + '.xml.gz';
    if (this.type) {
      file = this.type + '_' + file;
    }
    this.files.push(file);

    this.writerHandle = new SitemapXmlGen(path.join(this.filePath, file));
  }

  addLine(url, option) {
    if (this.lineIndex >= this.lineLimit) {
      this.lineIndex = 0;
      this.newXml();
    }

    const opt = Object.assign({
      // lastmod: this.lastmod,
      changefreq: this.changefreq,
      priority: this.priority,
      mobile: this.mobile
    }, option);

    this.lineIndex++;
    this.writerHandle.writeItem({
      loc: url,
      lastmod: opt.lastmod,
      changefreq: opt.changefreq,
      priority: opt.priority,
      mobile: opt.mobile
    });
  }

  end() {
    this.writerHandle && this.writerHandle.end();
    this.writerHandle = null;
  }

  getFiles() {
    return this.files;
  }
}

module.exports = SubSitemap;


lib/sitemap-xml-gen.js
/**
 * 扩展gzwriter服务，写入标准的xml服务，提供单行xml写入方法
 */

'use strict';

const GzWriter = require('./gz-writer');

class SitemapXmlGen {
  constructor(path) {
    this.outs = new GzWriter(path);
    this.outs.write('<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">\n');
  }

  /**
   * 单行写入
   * @param {object} data eq. {loc: 'http://xxx', lastmod: '2017-06-03'}
   */
  writeItem(data) {
    let strXml = '<url>\n';

    strXml += '<loc>' + data.loc + '</loc>\n';

    if (data.mobile) {
      strXml += '<mobile:mobile type="' + data.mobile + '"/>\n';
    }
    if (data.lastmod) {
      strXml += '<lastmod>' + data.lastmod + '</lastmod>\n';
    }
    if (data.changefreq) {
      strXml += '<changefreq>' + data.changefreq + '</changefreq>\n';
    }
    if (data.priority) {
      strXml += '<priority>' + data.priority + '</priority>\n';
    }
    strXml += '</url>\n';
    
    this.outs.write(strXml);
  }

  end(cb) {
    this.outs.write('</urlset>');
    this.outs.end(cb);
    this.outs = null;
  }
}

module.exports = SitemapXmlGen;
